torch_agents/cogment_verse_torch_agents/third_party/hive/agents/dqn.py

```diff
@@ -0,0 +1,321 @@
+            )
+
+        return {
+            "observation": update_info["observation"],
+            "action": update_info["action"],
+            "reward": update_info["reward"],
+            "done": update_info["done"],
+        }
+
+    def preprocess_update_batch(self, batch):
+        for key in batch:
+            batch[key] = torch.tensor(batch[key]).to(self._device)
+        return (batch["observation"],), (batch["next_observation"],)
+
+    @torch.no_grad()
+    def act(self, observation):
+        """Returns the action for the agent. If in training mode, follows an epsilon
+        greedy policy. Otherwise, returns the action with the highest q value."""
+
+        # Determine and log the value of epsilon
+        if self._training:
+            if not self._learn_schedule.get_value():
+                epsilon = 1.0
+            else:
+                epsilon = self._epsilon_schedule.update()
+            if self._logger.update_step(self._timescale):
+                self._logger.log_scalar("epsilon", epsilon, self._timescale)
+        else:
+            epsilon = self._test_epsilon
+
+        # Sample action. With epsilon probability choose random action,
+        # otherwise select the action with the highest q-value.
+        observation = (
+            torch.tensor(np.expand_dims(observation, axis=0)).to(self._device).float()
+        )
+        qvals = self._qnet(observation)
+        if self._rng.random() < epsilon:
+            action = self._rng.integers(self._act_dim)
+        else:
+            action = torch.argmax(qvals).item()
+
+        if (
+            self._training
+            and self._logger.should_log(self._timescale)
+            and self._state["episode_start"]
+        ):
+            self._logger.log_scalar("train_qval", torch.max(qvals), self._timescale)
+            self._state["episode_start"] = False
+        return action
+
+    def update(self, batch):
+        """
+        Updates the DQN agent.
+
+        Args:
+            update_info: dictionary containing all the necessary information to
+            update the agent. Should contain a full transition, with keys for
+            "observation", "action", "reward", and "done".
+        """
+        # if update_info["done"]:
+        #     self._state["episode_start"] = True
+
+        # if not self._training:
+        #     return
+
+        # Add the most recent transition to the replay buffer.
+        # self._replay_buffer.add(**self.preprocess_update_info(update_info))
+
+        # Update the q network based on a sample batch from the replay buffer.
+        # If the replay buffer doesn't have enough samples, catch the exception
+        # and move on.
+        if (
+            self._learn_schedule.update()
+            and self._replay_buffer.size() > 0
+            and self._update_period_schedule.update()
+        ):
+            # batch = self._replay_buffer.sample(batch_size=self._batch_size)
```

Comment:
I'm not sure why you changed that, you can definitely update the run impl to not own the replay buffer and just call update whenever you have a new sample.

Reply:
updated